src/api/StepsExchanger.ts

```diff
@@ -0,0 +1,135 @@
+ *
+ * Software distributed under the License is distributed on an “AS IS” basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the specific language governing rights and limitations under the License.
+ *
+ * The Original Code is manuscripts-frontend.
+ *
+ * The Original Developer is the Initial Developer. The Initial Developer of the Original Code is Atypon Systems LLC.
+ *
+ * All portions of the code written by Atypon Systems LLC are Copyright (c) 2022 Atypon Systems LLC. All Rights Reserved.
+ */
+import { CollabProvider } from '@manuscripts/body-editor'
+import { schema } from '@manuscripts/transform'
+import { Step } from 'prosemirror-transform'
+
+import Api from './Api'
+import { saveWithDebounce } from './savingUtilities'
+
+const MAX_ATTEMPTS = 20
+const THROTTLING_INTERVAL = 1200
+
+export type Subscription = (value: boolean) => void
+
+export class ObservableBoolean {
+  value = false
+  private subs: Subscription[] = []
+
+  setValue(value: boolean) {
+    this.value = value
+    this.subs.forEach((s) => s(value))
+  }
+
+  onChange(sub: Subscription) {
+    this.subs.push(sub)
+  }
+}
+
+export class StepsExchanger extends CollabProvider {
+  projectID: string
+  manuscriptID: string
+  api: Api
+
+  debounce: ReturnType<typeof saveWithDebounce>
+  flushImmediately?: () => void
+
+  isThrottling: ObservableBoolean
+  attempt = 0
+
+  constructor(
+    projectID: string,
+    manuscriptID: string,
+    currentVersion: number,
+    api: Api
+  ) {
+    super()
+    this.projectID = projectID
+    this.manuscriptID = manuscriptID
+    this.currentVersion = currentVersion
+    this.isThrottling = new ObservableBoolean()
+    this.debounce = saveWithDebounce()
+    this.api = api
+  }
+
+  async sendSteps(
+    version: number,
+    steps: Step[],
+    clientID: string,
+    flush = false
+  ) {
+    this.currentVersion = version
+    this.flushImmediately = this.debounce(
+      async () => {
+        const data = {
+          steps: steps.map((s) => s.toJSON()),
+          version,
+          clientID,
+        }
+        const response = await this.api.sendSteps(
+          this.projectID,
+          this.manuscriptID,
+          data
+        )
+        if (response.error === 'conflict' && this.attempt < MAX_ATTEMPTS) {
+          // debugger
+          console.warn('Retrying')
+          this.newStepsListener()
+          this.attempt++
+        } else if (response.error) {
+          console.error('Failed to send steps', response.error)
+        } else {
+          this.attempt = 0
+        }
+      },
+      THROTTLING_INTERVAL,
+      flush,
+      () => this.isThrottling.setValue(false)
+    )
+
+    if (!flush) {
+      this.isThrottling.setValue(true)
+    }
+
+    return Promise.resolve()
+  }
+
+  async receiveSteps(version: number, steps: unknown[], clientIDs: number[]) {
+    if (steps.length) {
+      //TODO send steps to listener
```

Comment:
we can probably remove that comment

Reply:
This is a comment I added so we don't forget to do this. The current body-editor api doesn't accept any args, and I didn't want to change it now.